silo-core/contracts/incentives/base/BaseIncentivesController.sol

```diff
@@ -0,0 +1,243 @@
+        if (_user == address(0)) revert InvalidUserAddress();
+        if (_to == address(0)) revert InvalidToAddress();
+
+        _;
+    }
+
+    constructor(address _owner, address _notifier) DistributionManager(_owner, _notifier) {}
+
+    /// @inheritdoc ISiloIncentivesController
+    function createIncentivesProgram(DistributionTypes.IncentivesProgramCreationInput memory _incentivesProgramInput)
+        external
+        onlyOwner
+    {
+        _createIncentiveProgram(_incentivesProgramInput);
+    }
+
+    /// @inheritdoc ISiloIncentivesController
+    function updateIncentivesProgram(
+        string calldata _incentivesProgram,
+        uint40 _distributionEnd,
+        uint104 _emissionPerSecond
+    ) external onlyOwner {
+        require(_distributionEnd >= block.timestamp, InvalidDistributionEnd());
+
+        bytes32 programId = getProgramId(_incentivesProgram);
+
+        require(_incentivesProgramIds.contains(programId), IncentivesProgramNotFound());
+
+        uint256 totalSupply = _shareToken().totalSupply();
+
+        _updateAssetStateInternal(programId, totalSupply);
+
+        incentivesPrograms[programId].distributionEnd = _distributionEnd;
+        incentivesPrograms[programId].emissionPerSecond = _emissionPerSecond;
+
+        emit IncentivesProgramUpdated(_incentivesProgram);
+    }
+
+    /// @inheritdoc ISiloIncentivesController
+    function handleAction(
+        bytes32 _incentivesProgramId,
+        address _user,
+        uint256 _totalSupply,
+        uint256 _userBalance
+    ) public onlyNotifier {
+        uint256 accruedRewards = _updateUserAssetInternal(_incentivesProgramId, _user, _userBalance, _totalSupply);
+
+        if (accruedRewards != 0) {
+            uint256 newUnclaimedRewards = _usersUnclaimedRewards[_user][_incentivesProgramId] + accruedRewards;
+            _usersUnclaimedRewards[_user][_incentivesProgramId] = newUnclaimedRewards;
+
+            emit RewardsAccrued(
+                _user,
+                incentivesPrograms[_incentivesProgramId].rewardToken,
+                getProgramName(_incentivesProgramId),
+                newUnclaimedRewards
+            );
+        }
+    }
+
+    /// @inheritdoc ISiloIncentivesController
+    function getRewardsBalance(address _user, string calldata _programName)
+        external
+        view
+        returns (uint256 unclaimedRewards)
+    {
+        bytes32 programId = getProgramId(_programName);
+
+        unclaimedRewards = _usersUnclaimedRewards[_user][programId];
+
+        (uint256 stakedByUser, uint256 totalStaked) = _getScaledUserBalanceAndSupply(_user);
+
+        unclaimedRewards += _getUnclaimedRewards(programId, _user, stakedByUser, totalStaked);
+    }
+
+    /// @inheritdoc ISiloIncentivesController
+    function claimRewards(address _to) external returns (AccruedRewards[] memory accruedRewards) {
+        if (_to == address(0)) revert InvalidToAddress();
+
+        accruedRewards = _accrueRewards(_to);
```

Comment:
Accrual of the rewards happens for `_to` but claiming happens for `msg.sender`.

Reply:
Done.

Reply:
Done